bodhi/tests/server/consumers/test_automatic_updates.py

```diff
@@ -0,0 +1,132 @@
+    """Test the automatic update handler."""
+
+    def setup_method(self, method):
+        """Set up environment for each test."""
+        super().setup_method(method)
+
+        self.release = self.create_release('47', create_automatic_updates=True)
+
+        body = {
+            'build_id': 442562,
+            'name': 'colord',
+            'tag_id': 214,
+            'instance': 's390',
+            'tag': 'f17-updates-candidate',
+            'user': 'sharkcz',
+            'version': '1.3.4',
+            'owner': 'sharkcz',
+            'release': '1.fc26',
+        }
+
+        self.sample_message = Message(topic='', body=body)
+        self.sample_nvr = f"{body['name']}-{body['version']}-{body['release']}"
+
+        self.db_factory = base.TransactionalSessionMaker(self.Session)
+        self.handler = AutomaticUpdateHandler(self.db_factory)
+
+    # Test the main code path.
+    def test_consume(self):
+        """Assert that messages about tagged builds create an update."""
+        # process the message
+        self.handler(self.sample_message)
+
+        # check if the update exists
+        update = self.db.query(Update).filter(
+            Update.builds.any(Build.nvr == self.sample_nvr)
+        ).first()
+        assert update is not None
+
+    # The following tests cover lesser-travelled code paths.
+
+    @mock.patch('bodhi.server.consumers.automatic_updates.transactional_session_maker')
+    def test___init___without_db_factory(self, transactional_session_maker):
+        """__init__() should create db_factory if missing."""
+        handler = AutomaticUpdateHandler()
+
+        assert handler.db_factory is transactional_session_maker.return_value
+        transactional_session_maker.assert_called_once_with()
+
+    # Test robustness: malformed messages, unknown koji builds, incomplete
+    # buildinfo, release missing from the DB
+
+    def test_missing_mandatory_elems(self):
+        """Test tag message without mandatory elements."""
+        msg = deepcopy(self.sample_message)
+        for elem in ('build_id', 'name', 'version', 'release'):
+            del msg.body[elem]
+        with pytest.raises(BodhiException):
+            self.handler(msg)
+
+    def test_unknown_koji_build(self):
+        """Test tag message about unknown koji build."""
+        msg = deepcopy(self.sample_message)
+        msg.body['release'] += '.youdontknowme'
+        with pytest.raises(BodhiException):
+            self.handler(msg)
+
+    def test_incomplete_koji_buildinfo(self):
+        """Test koji returning incomplete buildinfo."""
+        msg = deepcopy(self.sample_message)
+        msg.body['release'] += '.testmissingnvr'
+        with pytest.raises(BodhiException):
+            self.handler(msg)
+
+    # Test messages that should be ignored.
+
+    def test_ignored_tag(self, caplog):
+        """Test messages re: tags not configured for automatic updates."""
+        caplog.set_level(logging.DEBUG)
```

Comment:
The caplog fixture is nice :+1:

Reply:
But doesn't work with test classes derived from `unittest.TestCase`, took me a while to figure out...